### PR TITLE
change: Calculate key hash without cardano-cli in prepare-chain-params

### PR DESCRIPTION
**Pull request by AmbientTea**
# Description

* moves existing blake2b implementation from main-chain-folower-api to new `crypto` module in domain crate
* `prepare-chain-params` now uses it to calculate the key-hash of governance authority instead of externally calling cardano-cli

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages.
- [x] New tests are added if needed and existing tests are updated.
- [ ] Relevant logging and metrics added
- [ ] CI passes. See note on CI.
- [x] Any changes are noted in the `changelog.md` for affected crate
- [x] Self-reviewed the diff

# Note on CI
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges.  Please contact IOG Partner Chains developers to do this
for you.

### CODE REVIEW

File: primitives/domain/src/crypto.rs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+pub fn blake2b<const N: usize>(data: &[u8]) -> [u8; N] {
+	blake2b_simd::Params::new()
+		.hash_length(N)
+		.hash(data)
+		.as_bytes()
+		.try_into()
+		.expect("hash output always has expected length")
```
Comment (vvviiimmm) [inline primitives/domain/src/crypto.rs:7]: minor: can we say what the expected length is?

**Review by vvviiimmm — APPROVED** [review]

File: primitives/domain/src/lib.rs [hunk 2]
```diff
@@ -137,11 +141,14 @@ pub struct MainchainAddress(BoundedVec<u8, ConstU32<MAX_MAINCHAIN_ADDRESS_BYTES>
...
-	type Err = &'static str;
+	type Err = String;
 
 	fn from_str(s: &str) -> Result<Self, Self::Err> {
 		let bytes: Vec<u8> = s.as_bytes().to_vec();
-		let bounded = BoundedVec::try_from(bytes).map_err(|_| "Invalid length")?;
+		let len = bytes.len();
+		let bounded = BoundedVec::try_from(bytes).map_err(|_| {
+			format!("Invalid length: {len}, expected max {MAX_MAINCHAIN_ADDRESS_BYTES}")
+		})?;
```
Comment (LGLO) [inline primitives/domain/src/lib.rs:151]: How is this change related to hashing?

File: partner-chains-cli/src/prepare_configuration/prepare_chain_params.rs [hunk 3]
```diff
@@ -254,22 +295,19 @@ mod tests {
...
-			MockIO::eprint("Governance authority has been set to 0x76da17b2e3371ab7ca88ce0500441149f03cc5091009f99c99c080d9"),
-			MockIO::eprint(CHAIN_ID_PROMPT),
-			prompt_with_default_and_save_to_existing_file(CHAIN_ID, Some("1"),"2"),
-			silently_fill_legacy_chain_params(final_chain_config),
-		]);
+				scenarios::show_intro(),
+				MockIO::file_read(GOVERNANCE_AUTHORITY.config_file),
+				prompt_and_save_to_existing_file(CARDANO_PAYMENT_VERIFICATION_KEY_FILE, "payment.vkey"),
+				MockIO::file_read("payment.vkey"),
+				save_to_existing_file(GOVERNANCE_AUTHORITY, GOV_AUTH),
```
Comment (LGLO) [inline partner-chains-cli/src/prepare_configuration/prepare_chain_params.rs:305]: Cosmetic: it's a pity. GOV_AUTH is just short for GOVERNANCE_AUTHORITY, so basically the same name, but one is filename and one is key hex.

**Review by LGLO — APPROVED** [review]

File: primitives/domain/src/lib.rs (hunk 2 above)
Comment (AmbientTea) [reply primitives/domain/src/lib.rs:151]: It's not, I changed the wrong place and left it. Reverted.

File: partner-chains-cli/src/prepare_configuration/prepare_chain_params.rs (hunk 3 above)
Comment (AmbientTea) [reply partner-chains-cli/src/prepare_configuration/prepare_chain_params.rs:305]: Aliased the import and renamed the constant, it's unambiguous now.
